File: solutions/0796. Rotate String/0796_3.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
public:
    bool rotateString(const string &s, const string &goal) {
        if (s.size() != goal.size()) {
            return false;
        }
        size_t n = s.size();
        vector<size_t> prefix(1 + n, 0);
        size_t len = 0;
        for (size_t i = 1; i < n; ++i) {
            while (true) {
                if (goal[len] == s[i]) {
                    len++;
                    break;
                }
                if (len == 0) {
                    break;
                }
                len = prefix[len];
            }
            prefix[i + 1] = len;
        }
        len = 0;
        for (size_t i = 0; i < 2 * n; ++i) {
            while (true) {
                if (goal[len] == s[i % n]) {
                    len++;
                    break;
                }
                if (len == 0) {
                    break;
                }
                len = prefix[len];
            }
            if (len == n) {
                return true;
            }
        }
        return false;
    }
}; 

File: solutions/0796. Rotate String/0796_3.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        if len(s) != len(goal): 
            return False
        n = len(s)

        shifts = [1] * (n+1)
        left = -1
        for right in range(n):
            while left >= 0 and goal[left] != goal[right]:
                left -= shifts[left]
            shifts[right + 1] = right - left
            left += 1

        match_len = 0
        for char in s+s:
            while match_len >= 0 and goal[match_len] != char:
                match_len -= shifts[match_len]

            match_len += 1
            if match_len == n:
                return True

        return False  